packages/kit/types/index.d.ts

```diff
@@ -1071,8 +1071,8 @@ export type ActionResult<
 	Success extends Record<string, unknown> | undefined = Record<string, any>,
 	Invalid extends Record<string, unknown> | undefined = Record<string, any>
 > =
-	| { type: 'success'; status: number; data?: Success }
-	| { type: 'failure'; status: number; data?: Invalid }
+	| { type: 'success'; status: number; data: Success | void }
+	| { type: 'failure'; status: number; data: Invalid | void }
```

Comment:
I don't think we should do this. This makes the `data` attribute required if you construct this object yourself, which could be seen as a breaking change (besides being just incorrect that it's now required).

Reply:
When would you construct the object yourself? The docs are explicit that it's a shape SvelteKit generates...

> When calling a form action via fetch, the response will be one of these shapes.

...and right now the types are wrong because the fetched data will always include a `data` property, which can sometimes be `undefined`, in the `success` or `failure` case. If we don't want to change the types then we have to do something like this in `handle_action_json_request`...

```diff
+		/** @type {import('types').ActionResult} */
+		let result;

		if (data instanceof ActionFailure) {
-			return action_json({
+			result = {
				type: 'failure',
				status: data.status,
-				// @ts-expect-error we assign a string to what is supposed to be an object. That's ok
-				// because we don't use the object outside, and this way we have better code navigation
-				// through knowing where the related interface is used.
-				data: stringify_action_response(data.data, /** @type {string} */ (event.route.id))
-			});
			};
+
+			if (data.data) {
+				// @ts-expect-error we assign a string to what is supposed to be an object. That's ok
+				// because we don't use the object outside, and this way we have better code navigation
+				// through knowing where the related interface is used.
+				result.data = stringify_action_response(data.data, /** @type {string} */ (event.route.id))
+			}
		} else {
-			return action_json({
+			result = {
				type: 'success',
				status: data ? 200 : 204,
-				// @ts-expect-error see comment above
-				data: stringify_action_response(data, /** @type {string} */ (event.route.id))
-			});
+			};
+
+			if (data) {
+				result.data = stringify_action_response(data, /** @type {string} */ (event.route.id))
+			}
		}

+		return action_json(result);
```

...which doesn't feel great. Semantically I think it makes waaaay more sense for `success` and `failure` to always have a `data` property that's sometimes `undefined` rather than two different shapes depending on whether data was returned.

Reply:
They are not present when appearing in the client though, because `undefined` is removed while stringifying the JSON, so `data?: Success` is correct.